DirectProgramming/DPC++FPGA/ReferenceDesigns/anr/src/row_stencil.hpp

```diff
@@ -0,0 +1,151 @@
+//                    is called for every pixel to perform the actual
+//                    convolution. The function definition should be as follows:
+//
+//    OutT MyStencilFunction(int, int, ShiftReg<InT, filter_size>,
+//                           FunctionArgTypes...)
+//
+//                    The user can provide extra arguments to the callback by
+//                    using the FunctionArgTypes parameter pack.
+// FunctionArgTypes:  The user-provided type parameter pack of the arguments to
+//                    pass to the callback function.
+//
+//
+// FUNCTION ARGUMENTS
+// rows:            The number of rows in the image.
+// cols:            The number of columns in the image.
+//                  computed by the IP is rows*cols.
+// zero_val:        The 'zero' value for the stencil. This is used to pad
+//                  the columns of the image.
+// func:            The user-defined functor. This is a callback that is called
+//                  to perform the 1D convolution.
+// stencil_args...: The parameter pack of arguments to be passed to the
+//                  user-defined callback functor.
+//
+template <typename InType, typename OutType, typename IndexT, typename InPipe,
+          typename OutPipe, unsigned filter_size, unsigned parallel_cols,
+          typename StencilFunction, typename... FunctionArgTypes>
+void RowStencil(IndexT rows, IndexT cols, const InType zero_val,
+                StencilFunction func, FunctionArgTypes... stencil_args) {
+  // types coming into and out of the kernel from pipes, respectively
+  using InPipeT = DataBundle<InType, parallel_cols>;
+  using OutPipeT = DataBundle<OutType, parallel_cols>;
+
+  // validate the IndexT
+
+  // number of pixels to pad to the columns with
+  constexpr int kPaddingPixels = filter_size / 2;
+
+  // the size of the shift register to hold the window
+  constexpr int kShiftRegSize = filter_size + parallel_cols - 1;
+  constexpr IndexT kColThreshLow = kPaddingPixels;
+
+  // static asserts
+  static_assert(filter_size > 1);
+  static_assert(parallel_cols > 0);
+  static_assert(IsPow2(parallel_cols));
+  static_assert(std::is_integral_v<IndexT>);
+  static_assert(std::is_invocable_r_v<OutType, StencilFunction, int, int,
+                                      ShiftReg<InType, filter_size>,
+                                      FunctionArgTypes...>);
+
+  // constants
+  const IndexT col_thresh_high = cols + kPaddingPixels;
+  const IndexT padded_cols = PadColumns<IndexT, filter_size>(cols);
+  const IndexT col_loop_bound = padded_cols / parallel_cols;
+
+  // the shift register
+  [[intel::fpga_register]] ShiftReg<InType, kShiftRegSize> shifty_pixels;
+
+  // initialize the contents of the shift register
+  #pragma unroll
+  for (int i = 0; i < kShiftRegSize; i++) {
+    shifty_pixels[i] = zero_val;
+  }
+
+  // There is a bug where using loop_coalese here prevents the shift register
+  // ('shifty_pixels') from being inferred as shift register.

```

Comment:
Given the bug, do you not need to manually coalesce the loops?  If not, then why ask the compiler to coalesce them for you?